internal/plugin.go

```diff
@@ -1621,6 +1613,128 @@ func (p *Plugin) createPlacement(
 	return
 }
 
+// generateClusterSelector determines the type of input and creates a map of selectors to be used in either the
+// clusterSelector or labelSelector field
+func (p *Plugin) generateSelector(
+	resolvedSelectors map[string]interface{},
+	keys []string,
+) (map[string]interface{}, error) {
+	// Determine cluster selector type
+	var selectorType string
+	var usedSelectors []string
+	var legacy bool
+
+	for key := range resolvedSelectors {
```

Comment:
Could we rely on the typed structs in the apimachinery module? Here is a prototype of what I was thinking:

```go
func (p *Plugin) generateSelector(resolvedSelectors map[string]interface{}) (map[string]interface{}, error) {
	if resolvedSelectors == nil {
		return map[string]interface{}{"matchExpressions": []interface{}{}}, nil
	}

	resolvedSelectorsJSON, err := json.Marshal(resolvedSelectors)
	if err != nil {
		return nil, err
	}

	resolvedSelectorsLS := metav1.LabelSelector{}
	decoder := json.NewDecoder(bytes.NewReader(resolvedSelectorsJSON))
	decoder.DisallowUnknownFields()

	err = decoder.Decode(&resolvedSelectorsLS)
	if err != nil {
		resolvedSelectorsLS = metav1.LabelSelector{}

		// Check if it's a legacy selector
		for label, value := range resolvedSelectors {
			valueStr, ok := value.(string)
			if !ok {
				// TODO: Maybe an error message about the decode failing?
				return nil, fmt.Errorf(
					"legacy placement value has non-string type as value: %s",
					fmt.Sprint(resolvedSelectors[label]),
				)
			}

			lsReq := metav1.LabelSelectorRequirement{Key: label}

			if valueStr == "" {
				lsReq.Operator = metav1.LabelSelectorOpExists

			} else {
				lsReq.Operator = metav1.LabelSelectorOpIn
				lsReq.Values = []string{valueStr}
			}

			resolvedSelectorsLS.MatchExpressions = append(resolvedSelectorsLS.MatchExpressions, lsReq)
		}
	}

	if p.usingPlR && resolvedSelectorsLS.MatchLabels != nil {
		return nil, fmt.Errorf("placementRule does not support matchLabels")
	}

	resolved, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&resolvedSelectorsLS)
	if err != nil {
		// TODO: This should likely be a panic since it's a failure in generating the legacy label selector.
		panic(err)
	}

	return resolved, nil
}
```

Reply:
sure. I'll replace the selector generation logic with this

Reply:
@willkutler after thinking about this, if the JSON decode succeeds, maybe you can just return `resolvedSelectors` directly?

Reply:
@mprahl do you mean something like this? 
```
func (p *Plugin) generateSelector(
	resolvedSelectors map[string]interface{},
	keys []string,
) (map[string]interface{}, error) {
	if resolvedSelectors == nil {
		return map[string]interface{}{"matchExpressions": []interface{}{}}, nil
	}

	resolvedSelectorsJSON, err := json.Marshal(resolvedSelectors)
	if err != nil {
		return nil, err
	}

	resolvedSelectorsLS := metav1.LabelSelector{}
	decoder := json.NewDecoder(bytes.NewReader(resolvedSelectorsJSON))
	decoder.DisallowUnknownFields()

	err = decoder.Decode(&resolvedSelectorsLS)
	if err != nil {
		resolvedSelectorsLS = metav1.LabelSelector{}

		// Check if it's a legacy selector
		for label, value := range resolvedSelectors {
			valueStr, ok := value.(string)
			if !ok {
				return nil, fmt.Errorf(
					"matchExpressions has non-string type as value: %s",
					fmt.Sprint(resolvedSelectors[label]),
				)
			}

			lsReq := metav1.LabelSelectorRequirement{Key: label}

			if valueStr == "" {
				lsReq.Operator = metav1.LabelSelectorOpExists

			} else {
				lsReq.Operator = metav1.LabelSelectorOpIn
				lsReq.Values = []string{valueStr}
			}

			resolvedSelectorsLS.MatchExpressions = append(resolvedSelectorsLS.MatchExpressions, lsReq)
		}

		resolved, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&resolvedSelectorsLS)
		if err != nil {
			panic(err)
		}

		return resolved, nil
	}

	return resolvedSelectors, nil
}
```